File: BOJ/1000~/4963.cpp
#include <iostream>
#include <algorithm>
#include <cmath>
#include <string>
#include <string.h>
#include <vector>
#include <queue>
#define ll long long int
using namespace std;

int a, b, dd[55][55], visit[55][55], sum;
int dx[8] = {1, 1, 0, -1, -1, -1, 0, 1};
int dy[8] = {0, 1, 1, 1, 0, -1, -1, -1};

int bfs(int x, int y) {
	queue<pair<int, int>> q;

	q.push({ x,y });

	while (!q.empty()) {
		int nowx = q.front().first;
		int nowy = q.front().second;
		q.pop();
		if (visit[nowx][nowy] == 0) {
			visit[nowx][nowy] = 1;
			for (int i = 0; i < 8; i++) {
				int nextx = nowx + dx[i];
				int nexty = nowy + dy[i];
				if (nextx >= 0 && nextx < b && nexty >= 0 && nexty < a) {
					if (visit[nextx][nexty] == 0) q.push({ nextx,nexty });
				}
			}
		}
	}
	return 1;
}

int main() {
	std::ios::sync_with_stdio(false);
	cin.tie(NULL);
	while (cin >> a >> b) {
		if (a == 0 && b == 0) break;
		sum = 0;
		memset(dd, 0, sizeof(dd));
		memset(visit, 0, sizeof(visit));
		for (int i = 0; i < b; i++) {
			for (int j = 0; j < a; j++) {
				cin >> dd[i][j];
				if (dd[i][j] == 0) visit[i][j] = 1;
			}
		}
		for (int i = 0; i < b; i++) {
			for (int j = 0; j < a; j++) {
				if (visit[i][j] == 0) sum += bfs(i, j);
			}
		}
		cout << sum << '\n';
	}
	return 0;
}

File: BOJ/1000~/4963.java
import java.util.Scanner;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {
	static int dy[] = { 0, 0, 1, 1, 1, -1, -1, -1 };
	static int dx[] = { 1, -1, 0, 1, -1, 0, 1, -1 };
	static String s[];
	static int m, n;
	static int arr[][];
	static boolean visit[][];
	
	static public class Pair {
		private int a, b;

		Pair(int a, int b) {
			this.a = a;
			this.b = b;
		}

		public int first() {
			return a;
		}

		public int second() {
			return b;
		}
	}

	public static void main(String[] args) throws Throwable {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		while (true) {
			s = br.readLine().split(" ");
			m = Integer.parseInt(s[0]);
			n = Integer.parseInt(s[1]);

			if (n == 0 && m == 0)
				break;

			arr = new int[n][m];
			visit = new boolean[n][m];
			int res = 0;
			Queue<Pair> q = new LinkedList<Pair>();

			for (int i = 0; i < n; i++) {
				s = br.readLine().split(" ");
				for (int j = 0; j < m; j++) {
					arr[i][j] = Integer.parseInt(s[j]);
					if (arr[i][j] == 0)
						visit[i][j] = true;
				}
			}

			for (int i = 0; i < n; i++) {
				for (int j = 0; j < m; j++) {
					if (visit[i][j] == false) {
						q.add(new Pair(i, j));
						
						while (!q.isEmpty()) {
							Pair p = q.poll();
							int y = p.first();
							int x = p.second();
							if(!visit[y][x])
								visit[y][x] = true;

							for (int k = 0; k < 8; k++) {
								int nextY = y + dy[k];
								int nextX = x + dx[k];

								if (nextY >= 0 && nextY < n && nextX >= 0 && nextX < m) {
									if (arr[nextY][nextX] == 1 && visit[nextY][nextX] == false) {
										visit[nextY][nextX] = true;
										q.add(new Pair(nextY, nextX));
									}
								}
							}
						}
						res++;
					}
				}
			}
			
			System.out.println(res);
		}
	}

}
